### PR TITLE
Feature/apigateway

### PR DESCRIPTION
**Pull request by dmoverton**

### CODE REVIEW

File: example-apigateway/Setup.hs [hunk 1]
```diff
@@ -0,0 +1,2 @@
+import Distribution.Simple
```
Comment (koterpillar) [inline example-apigateway/Setup.hs]: You can deploy more than one function from a single Haskell project.

File: example-apigateway/package.json [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  "name": "serverless-haskell-apigw-example",
+  "version": "1.0.0",
+  "description": "Example of serverless-haskell with APIGateway",
+  "main": "index.js",
+  "directories": {
+    "test": "test"
+  },
+  "dependencies": {
+    "serverless": "^1.26.0",
+    "serverless-haskell": "file:///Users/doverton/ws/serverless-haskell/serverless-plugin"
```
Comment (koterpillar) [inline example-apigateway/package.json]: Put the relative path back here.

File: example-apigateway/Setup.hs (hunk 1 above)
Comment (dmoverton) [reply example-apigateway/Setup.hs]: I'm not sure I understand what change you want me to make.

Comment (koterpillar) [reply example-apigateway/Setup.hs]: You can put both example functions (the old one and the new API Gateway one) into the single `example-project` as two different (Stack) components.

**Comment by dmoverton** [comment]
I've combined the two example projects, as suggested.

File: src/AWSLambda/Events/APIGateway.hs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    <*> o .:? "queryStringParameters" .!= HashMap.empty
+    <*> o .:? "pathParameters" .!= HashMap.empty
+    <*> o .:? "stageVariables" .!= HashMap.empty
+    <*> o .: "requestContext"
+    <*> o .:? "body"
+
+
+$(makeLenses ''APIGatewayProxyRequest)
+
+-- | A Traversal to get the request body, if ther is one.
```
Comment (koterpillar) [inline src/AWSLambda/Events/APIGateway.hs]: Typo.

File: src/AWSLambda/Events/APIGateway.hs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+responseNotFound :: APIGatewayProxyResponse body
+responseNotFound =
+  APIGatewayProxyResponse
+  { _agprsStatusCode = 404
+  , _agprsHeaders = HashMap.empty
+  , _agprsBody = Nothing
+  }
+
+-- | Specialisation of lambdaMain for API Gateway
```
Comment (koterpillar) [inline src/AWSLambda/Events/APIGateway.hs]: Even if it's a specialisation, it needs a proper description first.

**Comment by dmoverton** [comment]
I've added more documentation on `apiGatewayMain`, as requested. I've also made some changes to `requestBody` and `responseBody` to make them more usable.

File: src/AWSLambda/Events/APIGateway.hs [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+Based on https://github.com/aws/aws-lambda-dotnet/tree/master/Libraries/src/Amazon.Lambda.APIGatewayEvents
+-}
+module AWSLambda.Events.APIGateway where
+
+import           Control.Lens
+import           Data.Aeson
+import           Data.Aeson.Casing       (aesonDrop, camelCase)
+import           Data.Aeson.TH           (deriveFromJSON)
+-- import           Data.CaseInsensitive (CI (..))
```
Comment (koterpillar) [inline src/AWSLambda/Events/APIGateway.hs]: If you aren't using CI, clean it up from the cabal file as well.

File: src/AWSLambda/Events/APIGateway.hs [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+responseBody :: Setter' (APIGatewayProxyResponse body) (Maybe body)
+responseBody = agprsBody . at () . mapping unTextValue
+
+responseBodyEmbedded :: Setter' (APIGatewayProxyResponse (Embedded body)) (Maybe body)
+responseBodyEmbedded = responseBody . mapping unEmbed
+
+responseBodyBinary :: Setter' (APIGatewayProxyResponse Base64) (Maybe ByteString)
+responseBodyBinary = responseBody . mapping _Base64
+
+-- | Process incoming events from @serverless-haskell@ using a provided
```
Comment (koterpillar) [inline src/AWSLambda/Events/APIGateway.hs]: "Process HTTP requests", perhaps?

File: src/AWSLambda/Events/APIGateway.hs [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+-- >   print $ request ^. requestBody
+-- >   pure $ responseOK & responseBodyEmbedded ?~ [1, 2, 3]
+--
+-- The type parameters @reqBody@ and @resBody@ represent the types of request and response body, respectively.
+-- The @FromText@ and @ToText@ contraints are required because these values come from string fields
+-- in the request and response JSON objects.
+-- To get direct access to the body string, use @Text@ as the parameter type.
+-- To treat the body as a stringified embedded JSON value, use @Embedded a@, where @a@ has the
+-- appropriate @FromJSON@ or @ToJSON@ instances.
+-- To treat the body as base 64 encoded binary use @Base64@.
```
Comment (koterpillar) [inline src/AWSLambda/Events/APIGateway.hs]: Perhaps add examples of defining `handler` for each of these cases, to reduce friction?

**Comment by dmoverton** [comment]
I'm going to merge this now, but I'll improve the documentation, as you suggest, when I get a chance.

**Pull request merged**